src/InfoViz/Native/ParallelCoordinates/index.js

```diff
@@ -811,83 +772,82 @@ function parallelCoordinate(publicAPI, model) {
-    if (model.provider[method]) {
-      model.subscriptions.push(model.provider[method](handleHoverBinUpdate));
-    }
-  });
 
-  if (model.provider.isA('DataUpdateProvider')) {
-    model.updateSubscription = model.provider.subscribeToDataUpdate(
-      'histogram2d',
-      data => {
-        Object.keys(data).forEach(xName => {
-          Object.keys(data[xName]).forEach(yName => {
-            const histoPayload = data[xName][yName][0];
-            if (histoPayload.role && histoPayload.role.type && histoPayload.role.type === 'alldata') {
-              model.provider.setHistogram2D(xName, yName, histoPayload);
-            }
-          });
-        });
-        publicAPI.render();
+  // Handle active field change, update axes
+  if (model.provider.isA('FieldProvider')) {
+    // Monitor any change
+    model.subscriptions.push(model.provider.onFieldChange(() => {
+      model.axes.updateAxes(model.provider.getActiveFieldNames().map(name =>
+        ({ name, range: model.provider.getField(name).range })
+      ));
+    }));
+    // Use initial state
+    model.axes.updateAxes(model.provider.getActiveFieldNames().map(name =>
+      ({ name, range: model.provider.getField(name).range })
+    ));
+  }
+
+  // Handle bin hovering
+  if (model.provider.onHoverBinChange) {
+    model.subscriptions.push(model.provider.onHoverBinChange(handleHoverBinUpdate));
+  }
+
+  if (model.provider.isA('Histogram2DProvider')) {
+    model.histogram2DDataSubscription = model.provider.subscribeToHistogram2D(
+      allBgHistogram2d => {
+        if (Object.keys(allBgHistogram2d).length > 1) {
+          model.allBgHistogram2dData = allBgHistogram2d;
+          publicAPI.render();
+        } else {
+          model.allBgHistogram2dData = null;
+        }
       },
-      model.axes.getAxesPairs(), { nbins: 32 });
-    model.subscriptions.push(model.updateSubscription);
+      model.axes.getAxesPairs(),
+      {
+        numberOfBins: model.numberOfBins,
+        partial: false,
+      }
+    );
+
+    model.subscriptions.push(model.axes.onAxisListChange(axisPairs => {
+      model.histogram2DDataSubscription.update(axisPairs);
+    }));
+
+    model.subscriptions.push(model.histogram2DDataSubscription);
   }
 
   if (model.provider.isA('SelectionProvider')) {
-    model.dataSubscription = model.provider.subscribeToDataSelection(
+    model.selectionDataSubscription = model.provider.subscribeToDataSelection(
       'histogram2d',
       data => {
         model.selectionData = data;
         if (model.provider.getAnnotation()) {
           model.axes.resetSelections(model.provider.getAnnotation().selection, false, model.provider.getAnnotation().score, scoreToColor);
+          if (data['##annotationGeneration##'] !== undefined) {
+            if (model.provider.getAnnotation().generation === data['##annotationGeneration##']) {
+              // render from selection data change (same generation)
+              publicAPI.render();
+            }
```

Comment:
Need comment:
the missing 'else' is the one case we don't want to render - when generation numbers don't match. Why?